qiskit/circuit/library/data_preparation/state_preparation.py

```diff
@@ -0,0 +1,502 @@
+        # the qubits are in the zero state)
+        if self._inverse is False:
+            initialize_instr = disentangling_circuit.to_instruction().inverse()
+        else:
+            initialize_instr = disentangling_circuit.to_instruction()
+
+        q = QuantumRegister(self.num_qubits, "q")
+        initialize_circuit = QuantumCircuit(q, name="init_def")
+        initialize_circuit.append(initialize_instr, q[:])
+
+        return initialize_circuit
+
+    def _get_num_qubits(self, num_qubits, params):
+        """Get number of qubits needed for state preparation"""
+        if isinstance(params, str):
+            num_qubits = len(params)
+        elif isinstance(params, int):
+            if num_qubits is None:
+                num_qubits = int(math.log2(params)) + 1
+        else:
+            num_qubits = math.log2(len(params))
+
+            # Check if param is a power of 2
+            if num_qubits == 0 or not num_qubits.is_integer():
+                raise QiskitError("Desired statevector length not a positive power of 2.")
+
+            # Check if probabilities (amplitudes squared) sum to 1
+            if not math.isclose(sum(np.absolute(params) ** 2), 1.0, abs_tol=_EPS):
+                raise QiskitError("Sum of amplitudes-squared does not equal one.")
+
+            num_qubits = int(num_qubits)
+        return num_qubits
+
+    def inverse(self):
+        """Return inverted StatePreparation"""
+        return StatePreparation(self._params_arg, inverse=not self._inverse, label=self._label)
+
+    def broadcast_arguments(self, qargs, cargs):
+        flat_qargs = [qarg for sublist in qargs for qarg in sublist]
+
+        if self.num_qubits != len(flat_qargs):
+            raise QiskitError(
+                "StatePreparation parameter vector has %d elements, therefore expects %s "
+                "qubits. However, %s were provided."
+                % (2**self.num_qubits, self.num_qubits, len(flat_qargs))
+            )
+        yield flat_qargs, []
+
+    def validate_parameter(self, parameter):
+        """StatePreparation instruction parameter can be str, int, float, and complex."""
+
+        # StatePreparation instruction parameter can be str
+        if isinstance(parameter, str):
+            if parameter in ["0", "1", "+", "-", "l", "r"]:
+                return parameter
+            raise CircuitError(
+                "invalid param label {} for instruction {}. Label should be "
+                "0, 1, +, -, l, or r ".format(type(parameter), self.name)
+            )
+
+        # StatePreparation instruction parameter can be int, float, and complex.
+        if isinstance(parameter, (int, float, complex)):
+            return complex(parameter)
+        elif isinstance(parameter, np.number):
+            return complex(parameter.item())
+        else:
+            raise CircuitError(f"invalid param type {type(parameter)} for instruction  {self.name}")
+
```

Comment:
This will fix calling `StatePreparation.repeat` 
```suggestion

    def _return_repeat(self, exponent: float) -> "Gate":
        return Gate(name=f"{self.name}*{exponent}", num_qubits=self.num_qubits, params=[])

```